include/revng-c/TypeNames/PTMLCTypeBuilder.h

```diff
@@ -0,0 +1,245 @@
+
+  PTMLCTypeBuilder(llvm::raw_ostream &OutputStream, bool GenerateTaglessPTML) :
+    PTMLCBuilder(/* GenerateTagLessPTML = */ GenerateTaglessPTML),
+    Out(std::make_unique<OutStream>(OutputStream,
+                                    DecompiledCCodeIndentation,
+                                    GenerateTaglessPTML)) {}
+
+  OutStream &out() { return *Out; }
+  void updateTheOutputStream(llvm::raw_ostream &OutputStream) {
+    Out = std::make_unique<OutStream>(OutputStream,
+                                      DecompiledCCodeIndentation,
+                                      isGenerateTagLessPTML());
+  }
+
+public:
+  /// Return a string containing the C type name of \a Type and
+  /// a (possibly empty) \a InstanceName.
+  tokenDefinition::types::TypeString
+  getNamedCInstance(const model::Type &Type,
+                    llvm::StringRef InstanceName,
+                    llvm::ArrayRef<std::string> AllowedActions = {},
+                    bool OmitInnerTypeName = false) const;
+
+  tokenDefinition::types::TypeString
+  getTypeName(const model::Type &Type) const {
+    return getNamedCInstance(Type, "");
+  }
+
+public:
+  /// Return the name of the array wrapper that wraps \a ArrayType
+  tokenDefinition::types::TypeString
+  getArrayWrapper(const model::ArrayType &ArrayType) const;
+
+  /// Return a string containing the C Type name of the return type of
+  /// \a FunctionType, and a (possibly empty) \a InstanceName.
+  /// \note If F returns more than one value, the name of the wrapping struct
+  ///       will be returned instead.
+  tokenDefinition::types::TypeString
+  getNamedInstanceOfReturnType(const model::TypeDefinition &FunctionType,
+                               llvm::StringRef InstanceName,
+                               bool IsDefinition) const;
+
+  tokenDefinition::types::TypeString
+  getReturnTypeName(const model::TypeDefinition &FunctionType,
+                    bool IsDefinition) const {
+    return getNamedInstanceOfReturnType(FunctionType, "", IsDefinition);
+  }
+
+public:
+  /// Print the function prototype (without any trailing ';') of \a Function.
+  /// \note If the return value or any of the arguments needs a wrapper, print
+  ///       it with the corresponding wrapper type. The definition of such
+  ///       wrappers should have already been printed before this function is
+  ///       called.
+  OutStream &printFunctionPrototype(const model::TypeDefinition &FunctionType,
+                                    const model::Function &Function,
+                                    bool SingleLine = false);
+  OutStream &printFunctionPrototype(const model::TypeDefinition &FunctionType,
+                                    const model::DynamicFunction &Function,
+                                    bool SingleLine = false);
+  OutStream &printFunctionPrototype(const model::TypeDefinition &FunctionType);
+
+  OutStream &printSegmentType(const model::Segment &Segment);
```

Comment:
I find weird that some methods return strings and don't affect the stream while other return a reference to `OutStream`.

I think that overall this is a symptom of the fact that these builders you've designed are not sure if they should manage the stream (which could be a file stream) themselves, or if they just want to build strings and then the serialization of everything to the stream is someone else's job.

If it's the builder who manages the stream (which seems the case, given `unique_ptr`), there shouldn't be any need to give around mutable references to the stream itself.

Did you think about this? What's the actual intent?

Reply:
Ah, yeah, this has nothing to do with ownership or whatnot, it's just to let the users to directly append something _after_ whatever a printer does, think
```cpp
  printFunctionPrototype(/* whatever */) << ";\n";
```

And the fact that the reference is non-const is because it's not a good practice to ever hold const references to a stream, since, in principle, the only useful thing you can do with it is to modify it.

Reply:
I see. So it's for letting others append stuff at the end of the stream that is managed by the builder.

This kinda makes sense, but before your refactoring it wasn't really clear if the builder actually owned the stream or not. So I was wary of any situation where passing reference around could further complicate the picture. After the part of the refactoring I've seen, things are much more encapsulated, and the ownership of the stream is clear, so giving mutable references around is not scary anymore.

Reply:
I've thought about it and I still think that ideally the builder should be the only one allowed to write stuff in the stream. If the goal is to let users append stuff at the end of the stream, let's just provide an `append` method that takes strings and puts them on the stream.

Let's take a clear stance on this and change the code accordingly. As it is know it's just that the builders aren't sure if they want to manage the stream or not, and it ends up like they only want to do it partially, which is awkward.

Reply:
Like, the `out()` method shouldn't even be needed.

Reply:
Wait, no, `out()` method is a separate issue. That's specifically because of the attempt to reuse the same builder with different streams when emitting functions one by one.

Reply:
I don't thing it's ever used outside of the builder. And if such a use sneaked in from somewhere, `append` or something would easily replace that.

Reply:
That being said, now that I think about it, append is definitely a good idea! Since we probably want to sanitize such extra input just in case.

Reply:
> Wait, no, out() method is a separate issue. That's specifically because of the attempt to reuse the same builder with different streams when emitting functions one by one.

Eh is that actually the case? I think that's the reason why you need `updateTheOutputStream`.
But `out()`? Not convinced.

AFAICT from `git grep "out()"` (which seems to me the only place where it's used) the reason why we need `out()` is to allow clients of the builders to append stuff directly at the end of the stream that is supposed to be managed by the builder.
And I see 2 problems with that:
- that would allow someone to append nasty stuff on the stream without guards (especially if the reference to `Out` escapes somewhere weird) like e.g. non-PTML stuff, unescaped things, potentially anything
- even if that string was valid PTML, we still have a problem if it wasn't generated by the same builder we're calling `out()` on, because now the builder has global persistent state (which is necessay and we can't avoid) that may alter the generation of certain strings

I don't think we can solve both in the scope of this PR, but we can definitely improve the situation.

I think that we should drop `out()` and replace it with 2 different methods:
- `appendRaw`, that takes in a raw string, and depending on the builder state being tagless or not, escape it in HTML
- `appendPTML`, that takes in a string that is expected to be a) a valid PTML string (perhaps we can check?), and b) it should have been created by the same builder instance of the builder you're calling `appendPTML` on (this can't be checked, but at least sticks out in review much more than `out() << whatever`).
It's good that this "need" for `out()` emerges in only one place, but that's not an excuse for settling on something that is still awkward. if condition b) is not true, shit happens, and that should be stated as a "contract" with the user in the comment that describe how to use it.

I'm not in love with the names, but you get the idea.
I'm also not sure `appendRaw` is necessary. If not `append` is enough.

I think `append` instead of `out()` is not perfect but has the following advantages
1. it makes clear already from the name that the builder is in charge of managing the stream
2. it makes it impossible to end up having a reference to the stream escape somewhere where someone can misuse it
3. if you look at current uses of `out()` most of them will become `B.append(B.getstuff())` which makes it a no-brainer to check locally that you're not breaking contract b)
4. for the remaining usages of `out()` we have auxiliary data structures (typically maps) that are populated with `Map[V] = B.getstuff()` and then in another place we do `B.append(Map.at(V))`. so it's not immediate to check for contract b), but as long as all the data injected into the `Map` is generated with `B` it's manually checkable in little time with high accuracy
5. items 3 and 4 seem to cover all uses (AFAICT) of `out()` today and migrate them to `append`. whenever someone will `append` something new, that will stick out in review, and we'll make sure there are no new nasty users, but at least we have a clear contract to check

How does that sound?

Am I missing something?

Reply:
First of all, I'm all for `out() << X;` -> `append(X);`.

That being said, your `appendRaw` sounds nasty. Mostly because _every_ string is valid PTML string by definition. Tags only carry additional information. But know that I think about it, you probably just meant `sanitizeAndAppend`. Where it's just making sure we don't emit stuff like `<`, `>`and so on. But that can be a bit tricky too, since somebody might try to pass tagged string through that and get confused why it breaks :thinking:

Either way, it feels like we're on the same page for the important bit. We can tweak the small stuff once we see how it ends up looking.

Reply:
> you probably just meant `sanitizeAndAppend`. Where it's just making sure we don't emit stuff like `<`, `>` and so on.

Exactly. And to be clear, it should allow passing strings that contain e.g. `<`, but then depending on the tagless mode it should rewrite them as html or not.

> But that can be a bit tricky too, since somebody might try to pass tagged string through that and get confused why it breaks 🤔

Yeah I can see that, and that's why I said that perhaps `appendRaw` (or `sanitizeAndAppend` or even better `escapeAndAppend`) can be left out if it's not really needed in practice.

But if it turns out we can't do without it, then yeah, it's not immune to breakage, but if you end up passing an already escaped string to an API that is called `escapeAndAppend` I don't think you should be surprised if it prints garbage.
If anything, the name is an alert for the reviewer (if not for the developer) that should ask themselves: "are we sure that on this call to `escapeAndAppend` we're not passing in a string that was already escaped?".

I think this is the best we can do without overblowing this PR.

Reply:
Done.